En la página https://matematicaj.blogspot.com/2020/03/Test-de-analogias-verbales-resuelto-con-claves-y-respuestas-pre-universidad-pdf.html se presentan algunas analogías de palabras. En el primer caso se pregunta soldado es a ejercito como:
1. Ratón a gato,
2. Llave a llavero
3. Paradoja a contradicción 
4. Pulga a perro
5. Tornillo a máquina.

Usaremos el álgebra lineal para resolver el problema. A principio de semestre vimos como  algoritmos (como bolsa de palabras, Word2vec, Gensim, GloVe) aplicado a un conjunto de documentos (corpus), [incrustan las palabras](https://unipython.com/como-desarrollar-embeddings-incrustaciones-de-palabras-con-gensim/) <!--(https://es.wikipedia.org/wiki/Word_embedding)--> en $\mathbb{R}^n$.  

Algunas de estas incrutaciones tiene propiedades semánticas, es decir que algunas operaciones vectoriales son congruentes con el significado de las palabras. Como en el famoso ejemplo  Rey - Hombre + Mujer -> Reina 

Para este ejercicio vamos a utilizar la librería `spacy` con una base de datos de noticias (news) pequeña (sm) en español (es) llamada `es_core_news_sm`

* Lo primero es obtener el vector $\vec{u}_0$  para la palabra soldado y el vector $\vec{v}_0$ para la palabra ejército. Los vectores no están ni en $\mathbb{R}^2$ ni en $\mathbb{R}^3$ sino en $\mathbb{R}^96$

* Luego se obtiene el vector  $\vec{w}_0 =  \vec{u}_0 - \vec{v}_0$ que va desde soldado hasta ejército.

* De manera similar se obtiene el vector para la primera repuesta. Es decir, el vector $\vec{w}_1$ que va de raton a gato.  Finalmente se calcula el coseno del ángulo $\cos(\alpha_1)$ entre el vector  $\vec{w}_0$ y $\vec{w}_1$.

* Se repite los mismo para cada par de palabras, obteniendo los vectores $\vec{w}_2$, $\vec{w}_3$, $\vec{w}_4$, $\vec{w}_5$. Posteriormente los cosenos  $\cos(\alpha_2)$, $\cos(\alpha_3)$, $\cos(\alpha_4)$, $\cos(\alpha_5)$, entre los vectores  $\vec{w}_0$ y el respectivo $\vec{w}_i$.

*  Finalmente se selecciona la respuesta con el mayor coseno del ángulo.

In [2]:
import spacy
spacy_es = spacy.load("es_core_news_sm")
soldado=spacy_es("soldado").vector
soldado

array([ 1.6616583 ,  2.2189255 ,  0.34981376,  1.6703401 ,  2.0867515 ,
        1.3361075 , -0.16638641,  1.7989249 , -0.3785099 , -0.9531485 ,
       -4.787784  ,  0.5524312 , -0.27699035, -4.69687   , -1.3153002 ,
       -2.1852336 ,  0.08720505, -1.7302126 ,  2.4323547 ,  1.7860228 ,
       -0.19702435, -2.0001698 , -1.3060259 , -2.4178617 , -0.43159366,
        1.8175896 , -0.21075806, -0.9220848 , -1.0492164 , -0.6611053 ,
       -1.0855255 , -0.8291515 ,  0.27188474,  0.56854105,  0.51144755,
        1.1975454 , -0.676649  ,  1.793778  , -2.2901528 ,  1.2468784 ,
       -1.7368109 ,  1.179375  ,  0.1729405 ,  2.7017672 , -0.74960184,
        4.0208335 ,  1.1929208 ,  1.8065389 ,  0.2725471 , -1.377909  ,
        1.8582463 ,  0.10081023, -4.811147  ,  3.0806613 ,  2.0560303 ,
       -1.5004798 ,  3.8858168 , -0.33191317,  0.515605  ,  0.6267386 ,
       -0.66777545,  3.5780149 ,  1.2418798 ,  1.9644146 , -3.0513065 ,
       -0.909126  , -3.0529184 ,  2.6292071 ,  1.071341  , -2.80

In [6]:
ejercito=spacy_es("ejercito").vector
ejercito

array([ 1.4411745 , -2.5533943 ,  0.05001923,  0.5438783 ,  3.2191896 ,
       -0.6444907 ,  1.6833391 ,  0.6963616 ,  1.0586026 , -1.9332976 ,
       -1.7777734 , -2.7146525 , -1.3105258 , -3.7120824 ,  0.7392074 ,
       -3.6514225 , -1.5230751 ,  0.58200836, -0.08182347, -0.3671136 ,
       -2.2644074 ,  1.1282105 , -0.2524311 , -2.9431238 , -0.23749608,
        2.0388002 ,  1.8218774 ,  0.5554329 , -3.332225  , -1.4221575 ,
        0.9868597 , -0.93209875,  0.05404449,  1.4221487 ,  1.7075851 ,
       -0.14718217, -0.7244853 ,  1.5032868 , -3.2873194 , -0.9774724 ,
       -1.4539995 , -0.23640147,  0.8068797 ,  0.40529698,  0.77754277,
        3.4201064 ,  0.9482086 ,  3.5556595 ,  0.97663105,  1.0857005 ,
        2.2996306 ,  0.25581706,  0.9635985 , -1.1386635 ,  1.9516218 ,
       -2.1825576 , -0.14389592, -0.6038519 , -0.7177818 ,  0.42310238,
        0.28719002,  1.9969208 ,  0.7748813 ,  1.1907499 , -0.30491388,
       -0.7016783 , -1.1902373 , -0.12971541, -0.01195528, -3.66

In [19]:
len(ejercito)

96

In [7]:
soldado_ejercito = soldado - ejercito
soldado_ejercito

array([ 0.22048378,  4.77232   ,  0.29979452,  1.1264617 , -1.1324382 ,
        1.9805982 , -1.8497255 ,  1.1025634 , -1.4371125 ,  0.98014915,
       -3.0100107 ,  3.2670836 ,  1.0335355 , -0.98478746, -2.0545077 ,
        1.4661889 ,  1.6102802 , -2.312221  ,  2.5141783 ,  2.1531363 ,
        2.067383  , -3.1283803 , -1.0535948 ,  0.5252621 , -0.19409758,
       -0.2212106 , -2.0326355 , -1.4775177 ,  2.2830086 ,  0.76105225,
       -2.0723853 ,  0.10294724,  0.21784025, -0.85360765, -1.1961375 ,
        1.3447275 ,  0.0478363 ,  0.2904911 ,  0.99716663,  2.224351  ,
       -0.2828114 ,  1.4157765 , -0.6339392 ,  2.2964702 , -1.5271447 ,
        0.6007271 ,  0.24471223, -1.7491206 , -0.7040839 , -2.4636095 ,
       -0.44138432, -0.15500683, -5.774746  ,  4.219325  ,  0.1044085 ,
        0.68207777,  4.0297127 ,  0.27193874,  1.2333868 ,  0.20363623,
       -0.9549655 ,  1.581094  ,  0.46699852,  0.7736647 , -2.7463927 ,
       -0.20744771, -1.8626812 ,  2.7589226 ,  1.0832963 ,  0.85

In [8]:
raton = spacy_es("raton").vector
gato = spacy_es("gato").vector
raton_gato = raton - gato
raton_gato

array([ 1.7292609 , -2.6920793 ,  1.6449865 ,  0.585894  ,  0.18355125,
        1.7529455 ,  0.26742268, -0.2818724 , -1.1140999 ,  0.27326632,
       -2.1986642 , -3.502504  , -1.55283   ,  0.1946752 , -2.2821047 ,
       -0.9666953 ,  3.4755416 ,  0.89788115, -2.9915497 ,  2.104562  ,
        0.12125504, -1.684021  , -1.5519907 ,  0.64072376,  2.1139739 ,
        0.35596192,  1.9436569 , -1.2219086 ,  1.4982193 , -0.43171203,
        0.09910274, -3.6442175 ,  1.2660626 , -0.5276778 , -1.2549467 ,
        3.666636  ,  4.7239256 ,  2.6916842 ,  3.5553339 , -0.19087106,
        1.1508683 ,  0.17247939,  2.99386   , -3.9594293 ,  3.5035145 ,
       -2.3001122 ,  2.6698284 ,  3.0638754 , -1.2746164 , -2.5987515 ,
       -0.54226136, -2.3511488 , -1.9398496 , -3.476222  ,  1.7357211 ,
       -2.6485043 , -4.9510717 , -5.5852194 ,  1.6283147 , -3.3416224 ,
        1.7301524 ,  2.3105087 ,  1.9140531 ,  2.272368  ,  0.47610486,
        1.6500486 , -0.35094178, -2.0509653 , -2.1018271 , -0.37

In [11]:
import numpy as np
# Los vectores obtenidos son arreglos de numpy
# El producto punto entre dos vectores se puede realizar con el método `dot`
# La norma de un vector se realiza con la función `np.linalg.norm`

In [12]:
cos_preg_rta1 = soldado_ejercito.dot(raton_gato) / (np.linalg.norm(soldado_ejercito) * np.linalg.norm(raton_gato))
cos_preg_rta1

-0.00087636773

In [14]:
llave = spacy_es("llave").vector
llavero = spacy_es("llavero").vector
llave_llavero = llave - llavero

cos_preg_rta2 = soldado_ejercito.dot(llave_llavero) / (np.linalg.norm(soldado_ejercito) * np.linalg.norm(llave_llavero))
cos_preg_rta2

0.21956891

In [15]:
paradoja = spacy_es("paradoja").vector
contradiccion = spacy_es("contradicción").vector
paradoja_contradiccion = paradoja - contradiccion

cos_preg_rta3 = soldado_ejercito.dot(paradoja_contradiccion) / (np.linalg.norm(soldado_ejercito) * np.linalg.norm(paradoja_contradiccion))
cos_preg_rta3

-0.022749668

In [16]:
pulga = spacy_es("pulga").vector
perro = spacy_es("perro").vector
pulga_perro = pulga - perro

cos_preg_rta4 = soldado_ejercito.dot(pulga_perro) / (np.linalg.norm(soldado_ejercito) * np.linalg.norm(pulga_perro))
cos_preg_rta4

0.103010654

In [18]:
tornillo = spacy_es("tornillo").vector
maquina = spacy_es("maquina").vector
tornillo_maquina = tornillo - maquina

cos_preg_rta5 = soldado_ejercito.dot(tornillo_maquina) / (np.linalg.norm(soldado_ejercito) * np.linalg.norm(tornillo_maquina))
cos_preg_rta5

-0.03248876

coseno del ángulo con soldado ejercito    
| palabra1 | palabra2    |        |
|----------|-------------|---------------|
|raton     |gato         |-0.00087636773 |
|llave     |llavero      | 0.21956891    |
|paradoja  |contradiccion|-0.022749668   |
|pulga     |perro        | 0.103010654   |
|tornillo  |maquina      |-0.03248876    |

La mejor respuesta es la que el coseno del ángulo sea mayor. En este caso sería llave-llavero, que evectivamente coresponde a la respuesta correcta. Ya que al igual que soldado ejérciro, ambas tienen una relación de elemento - conjunto. 


En el siguiente listado está el ejercicio que le corresponde de la página https://matematicaj.blogspot.com/2020/03/Test-de-analogias-verbales-resuelto-con-claves-y-respuestas-pre-universidad-pdf.html

|	Ejercicio	|	Código	|
|	----	|	-------	|
|	1	|	Ejemplo	|
|	2	|	20172005165	|
|	3	|	20181005165	|
|	4	|	20182005001	|
|	5	|	20182005130	|
|	6	|	20191005103	|
|	7	|	20191005154	|
|	8	|	20191005158	|
|	9	|	20192005014	|
|	10	|	20192005029	|
|	11	|	20192005041	|
|	12	|	20192005110	|
|	13	|	20192005144	|
|	14	|	20201005013	|
|	15	|	20201005027	|
|	16	|	20201005046	|
|	17	|	20201005101	|
|	18	|	20201005131	|
|	19	|	20201005134	|
|	20	|	20201005169	|
|	21	|	20201015147	|
|	22	|	20201015176	|
|	23	|	20202005044	|
|	24	|	20202005047	|
|	25	|	20202005050	|
|	26	|	20202005083	|
|	27	|	20202005095	|
|	28	|	20202005096	|
|	29	|	20202005098	|
|	30	|	20202005101	|
|	31	|	20202005111	|
|	32	|	20202005112	|
|	33	|	20202005115	|
|	34	|	20202005118	|
|	35	|	20202005146	|
|	36	|	20202015006	|
|	37	|	20202015013	|
|	38	|	20202015015	|
|	39	|	20202015019	|
|	40	|	20202015028	|
|	41	|	20202015134	|

Llene los datos solicitados en el respectivo formulario

